### PR TITLE
Add devcontainer

### LINKED ISSUE #488: Support for Visual Studio Code DevContainer
Provide a `.devcontainer` folder in the source repository so that it makes it easier for contributors to get a working environment on their development machine. This would be a docker file with all the pre-requisites defined in it that could just be spun up when required. This would mean that the version of python and other libraries could be kept in step with the repo.

**References**
<https://code.visualstudio.com/docs/remote/containers>

If you'd be willing to accept this, I'd be more than happy to create a PR for it.

### PR DESCRIPTION
**Pull request by EliiseS**
Adding the initial devcontainer mentioned in issue #488

If this looks good, I can also add onto the contribution docs to refer to the devcontainer.

Let me know if I'm missing anything or you'd like something else in it.

### CODE REVIEW

**Comment by marianan** [comment]
Thanks @EliiseS for your PR, we will follow up with comments and next steps by end of week.

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+FROM python:${PYTHON_VERSION}-buster
+
+# Dependencies versions
+ARG DOCKER_COMPOSE_VERSION="1.29.2"
+ARG COOKIECUTTER_VERSION="1.7.3"
+ARG AZURE_CLI_VERSION="2.24.0-1~buster"
+ARG NODE_VERSION="15.x"
+ARG DOTNET_VERSION="5.0"
+ARG JDK_VERIOSN="2:1.11-71"
```
Comment (marianan) [inline .devcontainer/Dockerfile:12]: >JDK_VERIOSN [](http://example.com/codeflow?start=4&length=11)

nit: please fix typo

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ARG MAVEN_VERIOSN="3.6.0-1"
```
Comment (marianan) [inline .devcontainer/Dockerfile:13]: >MAVEN_VERIOSN [](http://example.com/codeflow?start=4&length=13)

nit: please fix typo

File: .devcontainer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends build-essential && \
+    apt-get install -y python2.7-dev libffi-dev libssl-dev
```
Comment (marianan) [inline .devcontainer/Dockerfile:17]: >python2.7-dev [](http://example.com/codeflow?start=23&length=13)

python 2.7 is no longer supported in iotedgedev

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (EliiseS) [reply .devcontainer/Dockerfile:13]: Great catch, thanks a bunch :)

File: .devcontainer/Dockerfile (hunk 3 above)
Comment (EliiseS) [reply .devcontainer/Dockerfile:17]: Ah! I'll remove it from the contribution docs too then ;)

File: .devcontainer/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Dependencies versions
+ARG PYTHON_VERSION="3.7.10"
```
Comment (EliiseS) [inline .devcontainer/Dockerfile:2]: @marianan Any desired default python version?

File: .devcontainer/Dockerfile (hunk 3 above)
Comment (EliiseS) [reply .devcontainer/Dockerfile]: I've also updated the doc with devcontainer docs. Let me know if that's not OK

File: .devcontainer/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ARG DOTNET_VERSION="5.0"
+ARG JDK_VERSION="2:1.11-71"
+ARG MAVEN_VERSION="3.6.0-1"
+
+RUN apt-get update && \
+    apt-get install -y --no-install-recommends build-essential && \
+    apt-get install -y libffi-dev libssl-dev
+
+# Install python libraries
+RUN pip install cookiecutter==${COOKIECUTTER_VERSION} docker-compose==${DOCKER_COMPOSE_VERSION}
```
Comment (marianan) [inline .devcontainer/Dockerfile:20]: virtualenv

**Review by marianan — APPROVED** [review]
:shipit:

**Comment by jongio** [comment]
Awesome contrib!  Thank you @EliiseS

**Pull request merged**